### PR TITLE
Fix more types on EntityRepository and FilterCollection

### PR DESCRIPTION
**Pull request by derrabus**
Backports from #9515 and #9524

### CODE REVIEW

File: lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php [hunk 1]
```diff
@@ -3018,7 +3018,7 @@ protected function _storeAssociationMapping(array $assocMapping)
      * Registers a custom repository class for the entity class.
      *
      * @param string|null $repositoryClassName The class name of the custom mapper.
-     * @psalm-param class-string<EntityRepository>|null $repositoryClassName
+     * @psalm-param class-string<ObjectRepository>|null $repositoryClassName
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: :thinking: what makes you widen this type?

Comment (derrabus) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: Custom repositories don't have to extend `EntityRepository`. Is is sufficient if they implement `ObjectRepository`.

https://github.com/doctrine/orm/blob/193c3abf0e5f7a7e2699e7c7ac37c1e713d79ef8/lib/Doctrine/ORM/Exception/InvalidEntityRepository.php#L14

https://github.com/doctrine/orm/blob/193c3abf0e5f7a7e2699e7c7ac37c1e713d79ef8/lib/Doctrine/ORM/Configuration.php#L832

https://github.com/doctrine/orm/blob/193c3abf0e5f7a7e2699e7c7ac37c1e713d79ef8/lib/Doctrine/ORM/Repository/RepositoryFactory.php#L21

Comment (derrabus) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: Hmm, but this also means that the types on the entity manager are wrong. I'll remove the widening for now and try to straighten this out in a separate PR.

Comment (derrabus) [reply lib/Doctrine/ORM/Mapping/ClassMetadataInfo.php]: Moved to #9531.

**Review by greg0ire — APPROVED** [review]

**Pull request merged**